core/shared/src/main/scala/spire/algebra/Gcd.scala

```diff
@@ -0,0 +1,25 @@
+package spire.algebra
+
+import annotation.tailrec
+import scala.{specialized => sp}
+
+import spire.syntax.eq._
+import spire.syntax.euclideanRing._
+
+trait Gcd[@sp(Int, Long, Float, Double) A] extends Any {
+  def gcd(a: A, b: A): A
+  def lcm(a: A, b: A): A
+
+  @tailrec final def euclid(a: A, b: A)(implicit eq: Eq[A], er: EuclideanRing[A]): A =
+    if (b === er.zero) a else euclid(b, a % b)
+}
+
+object Gcd {
+  @inline final def apply[A](implicit ev: Gcd[A]): Gcd[A] = ev
+
+  implicit def fromEuclideanRing[@sp(Int, Long, Float, Double) A: EuclideanRing: Eq] =
```

Comment:
Here is the most recent fix somebody had for `Polynomial`'s gcd: #558

The default algorithm here works for integers, but not really rationals, where we want to find a common denominator first. In these cases, we override the definition of gcd to account for this.

I think, more generally, if we have this type class and it is open for people to implement/override themselves, perhaps for correctness or maybe just for speed, then having this default be implicit is dangerous, since it is super easy to silently replace a user's instance with this default.